text/0000-custom-model-classes.md

```diff
@@ -0,0 +1,225 @@
+- Start Date: 2019-05-09
+- Relevant Team(s): data 
+- RFC PR: (after opening the RFC PR, update this with a link to it and update the file name)
+- Tracking: (leave this empty)
+
+# Custom Model Class RFC
+
+## Summary
+
+This RFC is a follow-up RFC for #293 RecordData and replaces [https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor](https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor)
+
+Create a way for addons and user to define their own implemnentation of a model clas. Adds an `instantiateRecord` method to `Store` which would allow addons and ED itself to offer a clean replacement the default DS.Model with a custom Model class. 
+
+## Motivation
+
+- Allowing Addons to experiment with replacing the default DS.Model implementation ([https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor](https://github.com/runspired/rfcs/blob/ember-data-custom-records/text/0000-ember-data-model-factory-for.md#ember-data--modelfactoryfor))
+- Currently Ember Data conflates schema information and record implementation together as part of properties defined on `DS.Model`.
+DS.Model both defines the schema, that is the attributes and relationships present on a type, and the model implementation, that is the
+actual class which is exposed to the app containing those properties. This has several disadvantages:
+- it makes it hard to optimize the schema information, and ties to a definition of a JS class
+- it makes it hard to statically analyze the schema
+- it requires a 1-1 correspondence between the number of backing types in the system and the number of JS classes, resulting
+in an app sometimes having to ship a much higher number of classes than it otherwise would have.
+
+This RFC represents the first step in separating the record implementation from schema information.
+
+## Detailed design
+
+After the work in [Record Data Errors RFC](https://github.com/emberjs/rfcs/pull/465),  [Record Data State RFC](https://github.com/emberjs/rfcs/pull/463), and [Request State Service RFC](https://github.com/emberjs/rfcs/pull/466) we have enough coverage in public apis that we can implement the entire DS.Model class with just a few changes to Store APIs. 
```

Comment:
Do we have a proof of concept for this in eg something like ember-m3 to make sure we have actually provided enough API to make those use cases possible?

Reply:
Yup, have a spike of both default DS.Model and m3

Reply:
@igorT can we post the spike somewhere public as an example people reviewing it can look at to see what this looks like "in the real world"?

Reply:
Sure thing, I'm in the process of extracting the merged rfcs out of it as well which will make it more readable as well

Reply:
Was this addressed?

Reply:
https://github.com/hjdivad/ember-m3/pull/294 and https://github.com/emberjs/data/tree/igor/schema-temp